### PR TITLE
[6-0-stable] Backport Upgrade-safe URL-safe CSRF tokens #39076

### PR DESCRIPTION
**Pull request by kamipo**
This is the backport of #39076 to make apps upgrade-safe from Rails 5.2.5.

I've made a PR #41805 to make CSRF tokens forward compatible, that is enough for 6.0.x -> 6.0.next.

But for 5.2.4.x and 5.2.5. #39076 is required to make apps upgrade-safe both 5.2.4.x -> 5.2.next (skip 5.2.5) and 5.2.5 -> 5.2.next.

If this way is preferable, I'll merge this and backport to 5-2-stable (with making this Ruby 2.2 compatible).

### CODE REVIEW

**Review by rafaelfranca — APPROVED** [review]

**Comment by yevhene** [comment]
@rafaelfranca Please release it. We are about to upgrade to Rails 6.0, but stuck on 5.2 with this issue.

**Comment by GCorbel** [comment]
Let me know if I didn't understand correct. The fix has been merged in branch 6-0-stable but wasn't in release 6.0.3.7 so, If we want to upgrade from Rails 5.2 to Rails 6.0, we have to use 6-0-stable or wait for the next release, right ?

**Comment by yevhene** [comment]
@GCorbel As I understood 6.0.3.7 not included the fix because it is 4 level release. We need to wait for 6.0.4.

**Comment by zzak** [comment]
@yevhene @GCorbel In the mean time, have you tried out the [6-0-stable branch](https://github.com/rails/rails/tree/6-0-stable) and let us know if there's any issues before a release is cut?

**Comment by yevhene** [comment]
@zzak This feature works fine now in `6-0-stable` branch

**Comment by GCorbel** [comment]
@zzak I saw no issue with `6-0-stable` branch.

**Comment by zzak** [comment]
Thank you both for testing! I will leave it up to rails-core to decide when to release this 🙏

**Comment by GCorbel** [comment]
Rails 6.0.4 has been released with this commit in it. Thanks.

**Pull request merged**
